<launch>
  <arg name="circuit" />

  <arg name="car_name" default="car_1" />
  <arg name="origin_is_center_of_rear_axle" default="false" />
  <arg name="horizon" default="0.75" />
  <arg name="integration_step" default="0.04" />
  <arg name="strategy" default="dwa" />
  <arg name="use_sehs" default="false" />
  <arg name="planning_safety_margin" default="0.3" />
  <arg name="planning_frequency" default="25" />
  <arg name="min_lookahead" default="0.75" />
  <arg name="max_lookahead" default="1.75" />

  <group>
    <node pkg="racer" name="circuit_analysis" type="circuit_node" output="screen">
      <rosparam command="load" file="$(arg circuit)" />
      <param name="lookahead" value="3" />

      <param name="state_topic" value="/$(arg car_name)/state" />
      <param name="waypoints_topic" value="/$(arg car_name)/waypoints" />
      <param name="waypoints_visualization_topic" value="/$(arg car_name)/visualization/waypoints" />
    </node>

    <node pkg="racer" name="state" type="current_state_node" output="screen">
      <param name="frequency" value="50" />
      <param name="motor_rpm_topic" value="/$(arg car_name)/motor_rpm" />
      <param name="origin_is_center_of_rear_axle" value="$(arg origin_is_center_of_rear_axle)" />

      <param name="state_topic" value="/$(arg car_name)/state" />
      <param name="command_topic" value="/$(arg car_name)/commands" />
      <param name="motor_rpm_topic" value="/$(arg car_name)/motor_rpm" />

      <param name="map_frame_id" value="map" />
      <param name="base_link_frame_id" value="$(arg car_name)_base_link" />
      <param name="odom_frame_id" value="odom" />
    </node>

    <node pkg="racer" name="planning" type="planning_node">
      <param name="map_topic" value="/obstacles/costmap/costmap" />
      <param name="normal_frequency" value="$(arg planning_frequency)" />
      <param name="max_frequency" value="50" />

      <param name="use_sehs" value="$(arg use_sehs)" />

      <param name="throttle_levels" value="6" />
      <param name="steering_levels" value="17" />

      <param name="min_throttle" value="0.0" />
      <param name="max_throttle" value="1.0" />
      <param name="max_right" value="-1.0" />
      <param name="max_left" value="1.0" />
      <param name="safety_margin" value="$(arg planning_safety_margin)" />

      <param name="time_step_s" value="$(arg integration_step)" />

      <param name="map_topic" value="/obstacles/costmap/costmap" />
      <param name="state_topic" value="/$(arg car_name)/state" />
      <param name="waypoints_topic" value="/$(arg car_name)/waypoints" />
      <param name="trajectory_topic" value="/$(arg car_name)/trajectory" />
      <param name="inflated_map_topic" value="/$(arg car_name)/planner_map" />
    </node>

    <node pkg="racer" name="following" type="trajectory_following_node" output="screen">
      <param name="frequency" value="50" />

      <param name="strategy" value="$(arg strategy)" />

      <!-- Pure pursuit -->
      <param name="min_lookahead" value="$(arg min_lookahead)" />
      <param name="max_lookahead" value="$(arg max_lookahead)" />
      <!-- -->

      <!-- DWA -->
      <param name="distance_to_obstacle_weight" value="600.0" />
      <param name="position_weight" value="500.0" />
      <param name="heading_weight" value="250.0" />
      <param name="velocity_weight" value="250.0" />
      <param name="acceleration_weight" value="50.0" />

      <param name="prediction_horizon_s" value="$(arg horizon)" />
      <param name="integration_step_s" value="$(arg integration_step)" />
      <param name="throttle_levels" value="12" />
      <param name="steering_levels" value="33" />
      <param name="min_throttle" value="0.0" />
      <param name="max_throttle" value="1.0" />
      <param name="max_right" value="-1.0" />
      <param name="max_left" value="1.0" />
      <!-- -->

      <param name="map_topic" value="/obstacles/costmap/costmap" />
      <param name="state_topic" value="/$(arg car_name)/state" />
      <param name="waypoints_topic" value="/$(arg car_name)/waypoints" />
      <param name="trajectory_topic" value="/$(arg car_name)/trajectory" />
      <param name="twist_topic" value="/$(arg car_name)/commands" />
      <param name="ackermann_topic" value="/$(arg car_name)/command" />
    </node>

    <node name="visualize_action" pkg="racer" type="visualize_action" output="screen">
      <param name="prediction_horizon" value="$(arg horizon)" />
      <param name="time_step_s" value="$(arg integration_step)" />

      <param name="state_topic" value="/$(arg car_name)/state" />
      <param name="command_topic" value="/$(arg car_name)/commands" />
      <param name="visualization_topic" value="/$(arg car_name)/visualization/action" />
    </node>

    <node name="visualize_plan" pkg="racer" type="visualize_plan" output="screen">
      <param name="trajectory_topic" value="/$(arg car_name)/trajectory" />
      <param name="visualization_topic" value="/$(arg car_name)/visualization/trajectory" />
    </node>
  </group>
</launch>